### PR TITLE
Add  function EvaluateFunctionPointer() and test cases. 

### PR DESCRIPTION
**Pull request by ke-yu**
This function will expand parameter array and call function pointer at runtime.

### CODE REVIEW

**Comment by lukechurch** [comment]
Please rename "EvaluateFunctionPointer" to "Evaluate" in the future I want to be able to be able to overload the method with string, a path etc, rather than restricting it to an FP

File: Core/ProtoCore/Lang/BuiltInMethods.cs [hunk 1]
```diff
@@ -1167,6 +1169,17 @@ public BuiltInMethods(Core core)
...
                 }
             });
+
+            Methods.Add(new BuiltInMethod
+                {
+                    ID = MethodID.kEvaluateFunctionPointer,
+                    ReturnType = TypeSystem.BuildPrimitiveTypeObject(PrimitiveType.kTypeVar, true, Constants.kUndefinedRank),
+                    Parameters = new List<KeyValuePair<string, Type>>
+                    {
+                        new KeyValuePair<string, Type>("fptr", TypeSystem.BuildPrimitiveTypeObject(PrimitiveType.kTypeFunctionPointer, false, Constants.kUndefinedRank)),
```
Comment (lukechurch) [inline Core/ProtoCore/Lang/BuiltInMethods.cs]: FuncPtr not fptr

File: Tests/ProtoTest/Associative/BuiltinMethodsTest.cs [hunk 2]
```diff
@@ -554,6 +554,71 @@ public void BIM31_Sort_duplicate()
...
+{
+    return = x + y + z;
+}
+
+def foo(x, y)
+{
+    return = x * y;
+}
+
+t = foo;
```
Comment (lukechurch) [inline Tests/ProtoTest/Associative/BuiltinMethodsTest.cs]: do you have to do this assignment, or can you just write x = Evaluate(foo, param)?

you should be able to do the later, but I think we might have a bug that prevents it?

File: Tests/ProtoTest/Associative/BuiltinMethodsTest.cs [hunk 3]
```diff
@@ -554,6 +554,71 @@ public void BIM31_Sort_duplicate()
...
+param = { 2, 3, 4 };
+x = EvaluateFunctionPointer(t, param);
+t = bar;
+";           
+            ExecutionMirror mirror = thisTest.RunScriptSource(code);
+            thisTest.Verify("x", 24); 
+        }
```
Comment (lukechurch) [inline Tests/ProtoTest/Associative/BuiltinMethodsTest.cs:619]: This needs quite a bit more testing.

Please can you add:
1. A replicating test case funcs = {foo, bar}
2. A garbage collection verification case
3. The referenced function calling another function
4. The reference function recusing
5. the referenced function taking a function as an argument and execing it
6. A imperative block inside a function execed this way

We've seen failures in all these cases in the past. I don't have any reason to believe they won't work, but I'd like us to be sure before we start relying on this as a compile target.

Feel free to request TD assistance if needs be

**Comment by ke-yu** [comment]
Function renamed; add test cases for replication.

**Pull request merged**